Dependencies/xr/Source/OpenXR/SceneUnderstanding.cpp

```diff
@@ -71,500 +73,463 @@ SceneUnderstanding::SceneUnderstanding::InitOptions::InitOptions(
+            observedXRObjectIds.insert(xrObject.id);
+
+            if (m_objectIds.find(xrObject.id) == m_objectIds.end())
             {
-                if (m_babylonPlanes.count(key) == 0)
-                {
-                    const auto newPlane = std::make_shared<xr::System::Session::Frame::Plane>();
-                    m_babylonPlanes[key] = newPlane;
-                    m_babylonPlanesById[newPlane->ID] = newPlane;
-                }
-
-                auto& babylonPlane = *m_babylonPlanes[key];
-                babylonPlane.Center = XrPoseToBabylonPose(object->Pose);
-                babylonPlane.PolygonFormat = xr::PolygonFormat::XYZ;
-
-                // Note: Without a normal its unclear how to define the front/back of the plane
-                // This is currently an arbitrary winding order due to this lack of information
-                constexpr uint8_t VALUES_IN_POINT = 3;
-                constexpr uint8_t VALUES_IN_XYZ_QUAD = 12;
-                babylonPlane.Polygon.resize(VALUES_IN_XYZ_QUAD);
-                babylonPlane.Polygon[0] = -1.f * xrPlane.extent.width / 2.f;
-                babylonPlane.Polygon[1] = -1.f * xrPlane.extent.height / 2.f;
-                babylonPlane.Polygon[2] = 0;
-                babylonPlane.Polygon[3] = 1.f * xrPlane.extent.width / 2.f;
-                babylonPlane.Polygon[4] = -1.f * xrPlane.extent.height / 2.f;
-                babylonPlane.Polygon[5] = 0;
-                babylonPlane.Polygon[6] = 1.f * xrPlane.extent.width / 2.f;
-                babylonPlane.Polygon[7] = 1.f * xrPlane.extent.height / 2.f;
-                babylonPlane.Polygon[8] = 0;
-                babylonPlane.Polygon[9] = -1.f * xrPlane.extent.width / 2.f;
-                babylonPlane.Polygon[10] = 1.f * xrPlane.extent.height / 2.f;
-                babylonPlane.Polygon[11] = 0;
-                babylonPlane.PolygonSize = babylonPlane.Polygon.size() / VALUES_IN_POINT; // Number of points
-
-                babylonPlane.ParentSceneObjectID = objectKey;
-
-                args.Planes.push_back(babylonPlane);
+                auto object = std::make_unique<SceneObject>();
+                const auto objectId = object->ID;
+                m_objectIds[xrObject.id] = objectId;
+                m_objects[objectId] = std::move(object);
             }
+
+            const auto objectId = m_objectIds.at(xrObject.id);
+            m_updatedObjects.push_back(objectId);
+            auto& object = m_objects.at(objectId);
+            object->Type = c_objectTypeMap.at(xrObject.kind);
         }
 
-        args.UpdatedPlanes.clear();
-        for (const auto xrKey : m_updatedPlanes)
+        std::unordered_set<xr::su::SceneObject::Id> xrObjectIdsToRemove{};
+        for (const auto& [xrObjectId, objectId] : m_objectIds)
         {
-            if (m_babylonPlanes.count(xrKey) > 0)
+            if (observedXRObjectIds.find(xrObjectId) == observedXRObjectIds.end())

```

Comment:
i think this ones a little bit different because we are inserting into a different set for removal. The set here we are checking against may or may not contain the key.